### ISSUE TITLE
New chart proposition: kube-prometheus-stack-crds

### ISSUE THREAD

**Issue by desaintmartin** [prometheus-community/helm-charts#2612]
### Is your feature request related to a problem ?

- We have kube-prometheus-stack chart with prometheus-operator CRDs
- In the [helm documentation](https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#method-2-separate-charts), it is suggested to have a separate chart for CRD management. This is especially useful to automate upgrades of CRDs.

### Describe the solution you'd like.

- I propose to create a new `kube-prometheus-stack-crds` chart + related update script which only contains the CRDs. There are already three charts doing this:
  - ours: https://artifacthub.io/packages/helm/wiremind/kube-prometheus-stack-crds
  - https://artifacthub.io/packages/helm/morremeyer/kube-prometheus-stack-crds
  - and one based on prometheus-operator: https://artifacthub.io/packages/helm/appscode/prometheus-operator-crds

Would it be acceptable?

### Describe alternatives you've considered.

Maintaining it outside of prometheus-community.

**Comment by desaintmartin** [comment]
cc @morremeyer for additional discussion since you automated upgrades of the CRDs.

**Comment by monotek** [comment]
Will you maintain it?

**Comment by morremeyer** [comment]
@ekeih and I have created https://github.com/morremeyer/charts/tree/main/charts/kube-prometheus-stack-crds which is updated 100% automated.

This is achieved using:

* The [renovate](https://github.com/renovatebot/renovate) GitHub App do automatically trigger updates
* A workflow that gets the charts from this repository and updates the templates directory in the *kube-prometheus-stack-crds* chart
* The @morremeyer-bot to commit back to the repository (necessary in morremeyer/charts to trigger all pipelines since pushes with the GitHub actions' default token don't trigger other actions)

As I use kube-prometheus-stack in all my setups, I will maintain this for the for forseeable future.

If there is an interest in integrating that into this repository, I'm happy to discuss it - it could make maintenance easier since both charts could probably be released at the same time.

**Comment by jcpunk** [comment]
I'd also be interested in this chart.  It would make it much easier to install cilium via helm.

The current workflow:

1. install cilium
2. install kube-prometheus
3. enable cilium service monitor

Since you cannot enable the serviceMonitor without the CRDs and you cannot install kube-prometheus via helm without a working CNI, this gets a bit ugly.

This is a bit frustrating with config management tools.  In step 1, a set of chart values are set.  These get changed in step 3.  So enforcement of the settings gets a bit ugly.

**Comment by monotek** [comment]
@desaintmartin 

you might want to comaintain https://github.com/prometheus-community/helm-charts/pull/2697

maybe remove the annotation part to keep crds as thjey are.

**Comment by dacamposol** [comment]
@monotek why to take out the annotations?

The CRDs with the annotations are automatically generated with the attached script, so there are no maintainability reasons.

Also, I think they could be quite useful, as applying certain annotations to some resources is quite useful for some use-cases, and it doesn't affect the default logic.

**Comment by monotek** [comment]
these "usecases" are most of the time argocd workarounds which is kind of anoying, as i'd like to keep everything to be deployment agnostic.

argocd supports server side apply globally, so the annotation should not be needed.
see: https://argo-cd.readthedocs.io/en/latest/user-guide/sync-options/#server-side-apply

**Comment by dacamposol** [comment]
@monotek Argo indeed allows it to be applied globally, but that doesn't mean that someone would like to apply the same policy to the whole application.

If I were to create an umbrella Chart with `kube-prometheus-stack` and `kube-prometheus-stack-crds` with the annotations option I could define the following on my application; so I would have control on the sync policies of each of the Charts independently, even if they're deployed as a single Application.

```yaml
spec:
  source:
    helm:
      skipCrds: true
      values: |
        kube-prometheus-stack-crds:
          annotations:
            argocd.argoproj.io/sync-options: Replace=true
```

We have to keep in mind that Helm application Charts are used as dependencies as well, and I don't think it's bad to give the power to the developer of defining some custom parameters which don't affect in any way the default functionality of the resource.

**Comment by dacamposol** [comment]
> Merging #2697 would be great. In the future, would it be acceptable to have renovate push merge requests?

Do you mean to have Renovate checking the version of the `prometheus-operator` to know if it needs to refresh the CRDs?

That would be indeed really nice solution... I'm thinking into add us both as maintainers of that Chart, and then we can discuss on how to do that, as I don't know if we can configure a custom logic upon version change on Renovate, so I'd have to do some research in the matter.

**Comment by desaintmartin** [comment]
It's a good start to have the chart, then try to automate it somehow!

**Comment by Flasheh** [comment]
Would it make sense for this chart to follow the operator/crd version instead of it's own? I doubt the templating itself is going to see much changes and you could indicate those changes with semver labels maybe if the crds themselves have not changed.

It  might not be in line with other charts in this repo but in this case it would make identifying the required chart a lot clearer.

**Comment by desaintmartin** [comment]
My proposal in https://github.com/prometheus-community/helm-charts/pull/2922

We could follow the same increases than appVersion but one layer upper:
- 0.50.0 -> 0.50.1 would increase a minor version
- 0.50.0 -> 0.51.0 would increase a major version (since it's often breaking)
- 0.50.0 -> 0.50.0 but with internal, non breaking, non features changes would increase patch version


The reasoning is that we will change the chart outside of any upgrade, so IMHO it won't be possible to keep it synchronized.

**Comment by heresandyboy** [comment]
Hi, I am watching for the output of this.
In order to be able to use these new helm charts, I believe I would need to be able to get the exact CRDs helm chart version that matches the main prometheus chart version.

In fact I am using prometheus bundled with **[robusta-dev/robusta](https://github.com/robusta-dev/robusta).**

At the moment I am using a the charts supplied here **[https://artifacthub.io/packages/helm/morremeyer/kube-prometheus-stack-crds](https://artifacthub.io/packages/helm/morremeyer/kube-prometheus-stack-crds)**
simply because I can easily lookup the correct version to install/update prior to installing robusta like below:

How would I be able to do this when your crd charts are versioned properly?

Currently I have to pull and inspect the version of **kube-prometheus-stack** in the dependencies list of robusta, then install the matching version of the crds from morremeyer's **kube-prometheus-stack-crds** charts.

```bash
helm pull robusta/robusta --untar
version=$(helm dependency list robusta | grep "kube-prometheus-stack" | awk '{print $2}')
helm upgrade --install kube-prometheus-stack-crds morremeyer/kube-prometheus-stack-crds -n kube-system --version "$version"

````

**Comment by desaintmartin** [comment]
Hello,
This is indeed the biggest drawback regarding separating `version` from `appVersion` and I find myself in the same situation. Having a direct link to https://artifacthub.io/packages/helm/prometheus-community/prometheus-operator-crds is a... not that acceptable workaround. Note that `helm search hub prometheus-operator-crds` shows appVersion and could easily be scripted.

Needs:
- Ability to change things not related to `appVersion`
- Ability to know version from a quick glance

I've tried different things this week-end, and versions like `0.63.0-1` seem to be not valid in chart-testing (which checks for strict semver, and considers it as same version than `0.63.0`).

We still need to be able to do breaking changes on this chart without increasing `appVersion`. What we're trying on our side for some of our other CRDs is to sync versions one layer upper (if `appVersion` is `0.50.0`, `version` is `50.0.0`), allowing to do every-day patches without de-sync, and keeping breaking changes tied to a version increase (drawbacks: breaking changes can be hidden since it happens rarely so people won't look at changelog, and it is not that transparent).

What do you think?

**Comment by morremeyer** [comment]
@desaintmartin I don't see the issues you're seeing. At least with how I built the chart in morremeyer/charts, there are no changes except for when the CRDs change.

Especially in cases like this, I'm a proponent of doing the smallest possible thing, so everyone can understand it. 

"My" CRD chart doesn't have an appVersion because:

* it is not needed, keeping the version of kpm and kpm-crds in sync guarantees compatibility
* an appVersion for that chart would always be incorrect since one verison number cannot represent the versions of all components correctly

This is why https://github.com/morremeyer/charts/tree/main/charts/kube-prometheus-stack-crds does not have an appVersion.

@heresandyboy I think your problem would be solved by using "my" chart since it tracks the version of kube-prometheus-stack exactly and releases are 100% automated with just a few minutes to a few hours delay.

**Comment by dacamposol** [comment]
@morremeyer precisely I think that @heresandyboy is currently using "your" chart.

@desaintmartin maybe we should just pinpoint the Chart version to the Prometheus version, as the probability of having breaking changes on it is minimal, as we just want to replicate the CRDs from Prometheus with the possibility of introducing some custom annotations.

**Comment by heresandyboy** [comment]
@morremeyer, @dacamposol yes I am indeed using the morremeyer charts as they were the easiest to determine the version of the CRDs to install to match the version of kube-prometheus-stack used by robusta.

**Comment by desaintmartin** [comment]
OK, but now we are facing a dumb problem: we already released versions 1.0.0, 1.1.0 and 2.0.0. Un-publishing them may break existing installations now
Would it be acceptable to un-publish and "rollback" to 0.63.0?

**Comment by dacamposol** [comment]
> OK, but now we are facing a dumb problem: we already released versions 1.0.0, 1.1.0 and 2.0.0. Un-publishing them may break existing installations now Would it be acceptable to un-publish and "rollback" to 0.63.0?

Hmm... We could keep the following logic:

1. We start publishing regarding to the version of the `prometheus-operator`. Without unpublishing the previous ones.
2. Once the `prometheus-operator` launches a `1.0.0`, then we launch a notification about the possible breaking change and more information.

If we move already to the version of `prometheus-operator`, I don't think we have already enough public to provoke a major issue.

**Comment by rouke-broersma** [comment]
I think you should do at least one but better two major releases to notify users about this change before it happens. After sufficient time release the proper version unpublish **all** wrong versions.

If you don't do at least one major version indicating a breaking change, users will miss this and not receive updates. They will update kube prometheus stack, forget about crds, and it will work up until it doesn't.
